### ISSUE TITLE
is there a way to speed up the loading of tokenizer (with custom added tokens)?

### ISSUE THREAD

**Issue by mnskim** [huggingface/tokenizers#853]
Hello, Thank you for a great code release. 

I'm trying to add new tokens to a pre-trained bert tokenizer (fast tokenizer), and the add_tokens function makes this very easy to do. 
It seems though that after adding ~4000 additional tokens to the vocab, when I save this tokenizer and reload it from the saved path using AutoTokenizer.from_pretrained, it takes roughly 5 minutes to load the tokenizer.

I'm wondering if I made a mistake somewhere in the process, or if there's any method to make the loading faster. Regarding the latter, I can see that an "added_tokens.json" file has been saved - Since this gives me the tokens to add, maybe there is a way to reconfigure the tokenizer such that it doesn't have to deal with added_tokens.json?

Thank you!

**Comment by Narsil** [comment]
Hi @mnskim ,

Here is a reproducible script:

```python
from transformers import AutoTokenizer
import datetime


tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
print("Loaded")
tokens = []
for i in range(4000):
    tokens.append(f"[SPECIAL_{i}]")

tokenizer.add_tokens(tokens)
print("Add tokens")

tokenizer.save_pretrained("local")

print("Saved")

start = datetime.datetime.now()

tokenizer = AutoTokenizer.from_pretrained("local")
print("Load", datetime.datetime.now() - start)
```

Does take 2mn on my computer.

It takes 45s in the rust.

Investigating a little.

**Comment by Narsil** [comment]
All the time is spent recreating the `regex::RegexSet` and is linked to the way Addedtokens are currently added.

This is not the only issue when many tokens are added:

https://github.com/huggingface/tokenizers/issues/848

Since it seems to be a recurring problem expect some work on that front.
The solution seems to be creating a Trie instead of a RegexSet.

Maintaining 100% backward compatibility should be doable but will require some care.
PRs welcome !

**Comment by Narsil** [comment]
Closing as it should have been fixed by #871 (Not yet released but on master)
